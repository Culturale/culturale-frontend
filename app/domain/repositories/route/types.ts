export enum RouteState {
  FullySequenced = 'FULLY_SEQUENCED',
  NotFullySequenced = 'NOT_FULLY_SEQUENCED',
}
