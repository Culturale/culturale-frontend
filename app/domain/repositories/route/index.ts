export * from './route';
export * from './route.interface';
export * from './types';
