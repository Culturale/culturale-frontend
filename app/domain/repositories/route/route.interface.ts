import type { IStop } from '~/domain/entities';

import type { RouteState } from './types';

/**
 * Representation the route the driver has to follow
 *
 * @property route -> List of stops the driver has to do
 * @property routeState -> Indicates which is the state of the sequenced stops of the route
 */
export interface IRoute {
  route: IStop[];
  routeState: RouteState;

  /**
   * Updates the route with the provided list of stops
   * @public
   */
  setRoute: (route: IStop[]) => void;

  /**
   * Updates the route state with the provided value
   * @public
   */
  setRouteState: (state: RouteState) => void;

  /**
   * Returns the stop of the route with id
   * @public
   */
  getStop: (id: string) => IStop;

  /**
   * Returns true if the route is correctly sequenced
   * @readonly
   */
  readonly isRouteSequenced: boolean;

  /**
   * Returns all the stops that are completed
   * @readonly
   */
  readonly completedStops: IStop[];

  /**
   * Returns all the stops without being overlapped by geocode
   * @readonly
   */
  readonly visibleStops: IStop[];
}
