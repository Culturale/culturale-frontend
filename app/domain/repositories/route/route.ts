import { makeAutoObservable } from 'mobx';

import type { IStop } from '~/domain/entities/stop';

import type { IRoute } from './route.interface';
import { RouteState } from './types';

export class Route implements IRoute {
  public route: IStop[];
  public routeState: RouteState;

  constructor() {
    this.route = [];
    this.routeState = RouteState.FullySequenced;

    makeAutoObservable(this);
  }

  public setRoute(route: IStop[]) {
    this.route = route;
  }

  public setRouteState(state: RouteState) {
    this.routeState = state;
  }

  public getStop(id: string): IStop {
    return this.route.find((stop) => stop.stopId === id) || null;
  }

  get isRouteSequenced(): boolean {
    return this.routeState === RouteState.FullySequenced;
  }

  get visibleStops(): IStop[] {
    const stops = this.route;
    const visibleStops: IStop[] = [];

    for (const stop of stops) {
      const index = visibleStops.findIndex((s) => s.geoHash === stop.geoHash);

      if (index !== -1) {
        const stopWithSameGeoHash = visibleStops[index];
        const needsToBeReplaced =
          stopWithSameGeoHash.sequenceNumber > stop.sequenceNumber;

        if (needsToBeReplaced) {
          visibleStops[index] = stop;
        }
      } else {
        visibleStops.push(stop);
      }
    }

    return visibleStops;
  }

  get completedStops(): IStop[] {
    return this.route.filter((stop) => stop.isCompleted);
  }
}
