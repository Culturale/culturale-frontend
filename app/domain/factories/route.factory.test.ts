import { routeFactory } from './route.factory';

describe('routeFactory', function () {
  let instance;

  beforeEach(function () {
    instance = routeFactory({
      driver_id: 'test-driver-id',
      // isSequenced: true,
      stops: []
    });
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });
});
