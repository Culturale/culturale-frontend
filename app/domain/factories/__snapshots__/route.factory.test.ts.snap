// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routeFactory can be created 1`] = `
Route {
  "route": Array [],
  "routeState": "FULLY_SEQUENCED",
}
`;
